File: src/ros_study_py/ros_study_py/my_service_server.py
import rclpy
from rclpy.node import Node
from ros_study_msgs.srv import MySrv

class MyServiceServer(Node):
    def __init__(self):
        super().__init__('my_service_server')
        # 서비스 생성
        self.srv = self.create_service(MySrv, 'my_service', self.callback)
        self.get_logger().info('Service server ready')

    def callback(self, request, response):
        self.get_logger().info(f'Received request: num={request.num}')
        response.res = request.num * 2.0  # 예: 두 배로 반환
        return response

def main(args=None):
    rclpy.init(args=args)
    node = MyServiceServer()
    rclpy.spin(node)
    node.destroy_node()
    rclpy.shutdown()

if __name__ == '__main__':
    main()

File: src/ros_study_cpp/src/my_service_server.cpp
#include "rclcpp/rclcpp.hpp"
#include "ros_study_msgs/srv/my_srv.hpp"

using std::placeholders::_1;
using std::placeholders::_2;

class MyServiceServer : public rclcpp::Node
{
public:
  MyServiceServer() : Node("my_service_server")
  {
    // 서비스 생성
    service_ = this->create_service<ros_study_msgs::srv::MySrv>(
      "my_service", 
      std::bind(&MyServiceServer::handle_service, this, _1, _2)
    );
    RCLCPP_INFO(this->get_logger(), "Service server ready");
  }

private:
  // 서비스 콜백 함수
  void handle_service(
    const std::shared_ptr<ros_study_msgs::srv::MySrv::Request> request,
    std::shared_ptr<ros_study_msgs::srv::MySrv::Response> response)
  {
    RCLCPP_INFO(this->get_logger(), "Received request: %f", request->num);
    response->res = request->num * 2.0f;  // 예: 두 배로 반환
  }

  // 서비스 핸들러 객체
  rclcpp::Service<ros_study_msgs::srv::MySrv>::SharedPtr service_;
};


int main(int argc, char **argv)
{
  rclcpp::init(argc, argv);
  rclcpp::spin(std::make_shared<MyServiceServer>());
  rclcpp::shutdown();
  return 0;
}
